File: backend/app/schemas/analytics.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List, Dict, Any

class PerformanceMetrics(BaseModel):
    sharpe_ratio: float
    max_drawdown: float
    win_rate: float
    total_trades: int
    total_pnl: float
    start_date: Optional[datetime]
    end_date: Optional[datetime]

class CorrelationRequest(BaseModel):
    symbols: List[str]
    timeframe: str = "1h"
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None

class CointegrationPair(BaseModel):
    asset_a: str
    asset_b: str
    score: float
    p_value: float
    is_cointegrated: bool
    z_score: float

class CorrelationResponse(BaseModel):
    matrix: Dict[str, Dict[str, float]]
    lead_lag_matrix: Dict[str, Dict[str, int]]
    cointegrated_pairs: List[CointegrationPair]

class RollingCorrelationPoint(BaseModel):
    time: str
    value: float

File: frontend/src/services/analytics.ts
import apiClient from './client';
import { CointegratedPair } from '@/types';

export interface PerformanceMetrics {
    sharpe_ratio: number;
    max_drawdown: number;
    win_rate: number;
    total_trades: number;
    total_pnl: number;
    start_date?: string;
    end_date?: string;
}

export interface CorrelationResponse {
    matrix: Record<string, Record<string, number>>;
    lead_lag_matrix?: Record<string, Record<string, number>>;
    cointegrated_pairs: {
        asset_a: string;
        asset_b: string;
        score: number;
        p_value: number;
        is_cointegrated: boolean;
        z_score: number;
    }[];
}

export const getPerformanceMetrics = async (startDate?: string, endDate?: string): Promise<PerformanceMetrics> => {
    const params: any = {};
    if (startDate) params.start_date = startDate;
    if (endDate) params.end_date = endDate;

    const response = await apiClient.get('/analytics/performance', { params });
    return response.data;
};

export const fetchCorrelationMatrix = async (symbols: string[], timeframe: string = '1h'): Promise<CorrelationResponse> => {
    const response = await apiClient.post<CorrelationResponse>('/analytics/correlation-matrix', { symbols, timeframe });
    return response.data;
};

export interface RollingCorrelationPoint {
    time: string;
    value: number;
}

export const fetchRollingCorrelation = async (symbol_a: string, symbol_b: string, timeframe: string = '1h', window: number = 30): Promise<RollingCorrelationPoint[]> => {
    const response = await apiClient.get<RollingCorrelationPoint[]>('/analytics/correlation/rolling', {
        params: { symbol_a, symbol_b, timeframe, window }
    });
    return response.data;
};
